pkg/3scale/amp/operator/apicast_reconciler.go

```diff
@@ -151,8 +161,112 @@ func (r *ApicastReconciler) apicastLogLevelEnvVarMutator(desired, existing *apps
+		idx := helper.FindVolumeMountByName(existingContainer.VolumeMounts, volumeNameToDelete)
+		if idx >= 0 {
+			// Found a existing volume that needs to be removed
+			// remove index
+			existingContainer.VolumeMounts = append(existingContainer.VolumeMounts[:idx], existingContainer.VolumeMounts[idx+1:]...)
+			changed = true
+		}
+	}
+
+	return changed
+}
+
+// volumeMountsMutator implements basic VolumeMount reconcilliation
+// Added when in desired and not in existing
+// Updated when in desired and in existing but not equal
+// Existing not in desired will NOT be removed. Allows manually added volumemounts
+func apicastVolumesMutator(desired, existing *appsv1.DeploymentConfig) bool {
+	changed := false
+	existingSpec := &existing.Spec.Template.Spec
+	desiredSpec := &desired.Spec.Template.Spec
+
+	// Add desired not in existing
+	for desiredIdx := range desiredSpec.Volumes {
+		existingIdx := helper.FindVolumeByName(existingSpec.Volumes, desiredSpec.Volumes[desiredIdx].Name)
+		if existingIdx < 0 {
+			existingSpec.Volumes = append(existingSpec.Volumes, desiredSpec.Volumes[desiredIdx])
+			changed = true
+		} else if !helper.VolumeFromSecretEqual(existingSpec.Volumes[existingIdx], desiredSpec.Volumes[desiredIdx]) {
+			existingSpec.Volumes[existingIdx] = desiredSpec.Volumes[desiredIdx]
+			changed = true
+		}
+	}
+
+	// Check custom policy annotations in existing and not in desired to delete volumes associated
+	// From the APIManager CR, operator does not know which custom policies have been deleted to reconcile volumes
+	// Only volumes associated to custom policies are deleted. The operator still allows manually arbitrary mounted volumes
+	existingCustomPolicyVolumeNames := component.ApicastVolumeNamesFromAnnotations(existing.Annotations)
+	desiredCustomPolicyVolumeNames := component.ApicastVolumeNamesFromAnnotations(desired.Annotations)
+	volumesToDelete := helper.ArrayStringDifference(existingCustomPolicyVolumeNames, desiredCustomPolicyVolumeNames)
+	for _, volumeNameToDelete := range volumesToDelete {
+		idx := helper.FindVolumeByName(existingSpec.Volumes, volumeNameToDelete)
+		if idx >= 0 {
+			// Found a existing volume that needs to be removed
+			// remove index
+			existingSpec.Volumes = append(existingSpec.Volumes[:idx], existingSpec.Volumes[idx+1:]...)
+			changed = true
+		}
+	}
+
+	return changed
+}
+
+func apicastCustomPolicyAnnotationsMutator(desired, existing *appsv1.DeploymentConfig) bool {
+	// It is expected that APIManagerMutator has already added desired annotations to the existing annotations
+	// find existing custom policy annotations not in desired and delete them
+	updated := false
+	existingCustomPolicyVolumeNames := component.ApicastVolumeNamesFromAnnotations(existing.Annotations)
+	desiredCustomPolicyVolumeNames := component.ApicastVolumeNamesFromAnnotations(desired.Annotations)
+	if !helper.StringSliceEqualWithoutOrder(existingCustomPolicyVolumeNames, desiredCustomPolicyVolumeNames) {
```

Comment:
Instead of doing the "string slice equal without order" wouldn't sorting the strings work too?

Reply:
It will also work, sort and then reflect.DeepEqual. That is another implementation to compare slices regardless order.